osbuild/util/toml.py

```diff
@@ -0,0 +1,78 @@
+"""
+Utility functions for reading and writing toml files.
+
+Handles module imports for all supported versions (in a build root or on a host).
+"""
+import importlib
+from types import ModuleType
+from typing import Optional
+
+# Different modules require different file mode (text vs binary)
+_toml_modules = {
+    "tomllib": {"mode": "rb"},  # stdlib since 3.11 (read-only)
+    "tomli": {"mode": "rb"},  # EL9+
+    "toml": {"mode": "r", "encoding": "utf-8"},  # older unmaintained lib, needed for backwards compatibility
+    "pytoml": {"mode": "r", "encoding": "utf-8"},  # deprecated, needed for backwards compatibility (EL8 manifests)
+}
+
+
+_toml: Optional[ModuleType] = None
```

Comment:
I played around a bit but the result does not really feel nicer then what we already have here (and for write it gets even worse), so I'm just listing it for completion. I wish python had better support for higher order functions, I think with lambdas that support "with" this coudld be written in a much nicer way. Oh well :)
```python
@contextlib.contextmanager
def _opener_rb(path):
    with open(path, "rb") as fp:
        yield fp


@contextlib.contextmanager
def _opener_null(path):
    yield path


def load_from_file(path):
    with _opener(path) as tomlfile:  # pylint: disable=unspecified-encoding
        return _toml.load(tomlfile)


# Different modules require different file mode (text vs binary)
_toml_modules = {
    "tomllib": _opener_rb,  # stdlib since 3.11 (read-only)
    "tomli": _opener_rb,    # EL9+
    "toml": _opener_null,   # older unmaintained lib, needed for backwards compatibility
    "pytoml": _opener_null, # EL8
}


_toml: Optional[ModuleType] = None
_opener: None
for module, opener in _toml_modules.items():
    try:
        _toml = importlib.import_module(module)
        _opener = opener
        break
    except ModuleNotFoundError:
        pass
else:
    raise ModuleNotFoundError("No toml module found: " + ", ".join(_toml_modules))
```

Reply:
Indeed, I don't think there's a clean way of doing this.  I considered something similar, having opener functions instead of `**kwargs` that get passed around.  I'm fine either way, really.